### PR TITLE
Caster wheel and ros_control in gazebo

### LINKED ISSUE #102: Free caster in the simulation via the description argument
## Desired behavior
Enable the use of a continuous joint in the simulation

## Additional context

Here https://github.com/Ekumen-OS/andino/pull/100 the option was added

### PR DESCRIPTION
**Pull request by olmerg**
# 🦟 Bug fix

Fixes #102

## Summary

This PR solve the issue of caster wheel and has it has an initial launch to use ros control in gazebo with the andino, but has a Bug with the model which do not let to move in a straight line.

**I submit for discussion**

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if it affects the public API)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: andino_gazebo/launch/andino_with_ros_control.launch.py
Comment (francocipollone) [inline andino_gazebo/launch/andino_with_ros_control.launch.py:1]: I have a general doubt, why the ros control configuration isn't happening in the spawn robot launch file? Maybe I am missing something

**Review by francocipollone — CHANGES_REQUESTED** [review]
Good job! Minor comments

File: andino_gazebo/launch/andino_with_ros_control.launch.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        cmd=[
+            'ros2',
+            'control',
+            'load_controller',
+            '--set-state',
+            'active',
+            'joint_state_broadcaster',
+        ],
+        output='screen',
+    )
```
[the remark this answers is not in the document]
Comment (olmerg) [reply andino_gazebo/launch/andino_with_ros_control.launch.py:78]: I following this approach [ros2 control examples](https://github.com/ros-controls/gazebo_ros2_control/blob/master/gazebo_ros2_control_demos/launch/diff_drive.launch.py)

File: andino_gazebo/launch/andino_with_ros_control.launch.py
Comment (olmerg) [reply andino_gazebo/launch/andino_with_ros_control.launch.py:1]: sounds great idea, I am going to try

File: andino_gazebo/launch/andino_with_ros_control.launch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (olmerg) [reply andino_gazebo/launch/andino_with_ros_control.launch.py:78]: In the other way , when I start controller Manager an error is generated 

```
[ros2_control_node-5] terminate called after throwing an instance of 'pluginlib::LibraryLoadException'
[ros2_control_node-5]   what():  According to the loaded plugin descriptions the class gazebo_ros2_control/GazeboSystem with base class type hardware_interface::SystemInterface does not exist. Declared types are  andino_base/DiffDriveAndino fake_components/GenericSystem mock_components/GenericSystem test_hardware_components/TestSystemCommandModes test_hardware_components/TestTwoJointSystem test_system
```

[the remark this answers is not in the document]
Comment (francocipollone) [reply andino_gazebo/launch/andino_with_ros_control.launch.py:78]: See this alternative: https://github.com/Ekumen-OS/andino/blob/humble/andino_control/launch/andino_control.launch.py#L61-L65

```
    diff_drive_controller_spawner = Node(
        package="controller_manager",
        executable="spawner",
        arguments=["diff_controller", "--controller-manager", "/controller_manager"],
    )
```

Using the cli isn't bad however better to treat it as a node in your launch file rather than an outsider process

File: andino_description/urdf/include/andino_caster_macro.urdf.xacro [hunk 2]
```diff
@@ -115,7 +115,8 @@
       </joint>
     </xacro:if>
     <xacro:unless value="${use_fixed}">
-      <joint name="caster_rotation_joint" type="continuous">
+      <!--TODO(olmerg) initial solution problem with ros control  -->
+      <joint name="caster_rotation_joint" type="fixed">
```
Comment (francocipollone) [inline andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: Not seeing why this change. This section of the if-unless clause is to have continuous joint.

In case you need to use the fixed one in your simulation pass a different argument to the xacro.

File: andino_gazebo/launch/andino_one_robot.launch.py [hunk 3]
```diff
@@ -67,7 +67,7 @@ def generate_launch_description():
         PythonLaunchDescriptionSource(
             os.path.join(pkg_andino_gazebo, 'launch', 'spawn_robot.launch.py')
         ),
-        launch_arguments={'use_gazebo_ros_control': 'false'}.items(),
+        # launch_arguments={'use_gazebo_ros_control': 'false'}.items(),
```
Comment (francocipollone) [inline andino_gazebo/launch/andino_one_robot.launch.py:70]: Is it ok to have it commented out?

File: andino_gazebo/urdf/andino.gazebo.xacro [hunk 4]
```diff
@@ -44,6 +47,15 @@
...
           <parameters>$(find andino_control)/config/andino_controllers.yaml</parameters>
       </plugin>
+      <plugin name="andino_joint_state_caster" filename="libgazebo_ros_joint_state_publisher.so">
+      <ros>
+        <remapping>~/out:=joint_states</remapping>
+        <remapping>/tf:=tf</remapping>
+      </ros>
+      <update_rate>10</update_rate>
+      <joint_name>caster_wheel_joint</joint_name>
+      <!-- <joint_name>caster_rotation_joint</joint_name> -->
```
Comment (francocipollone) [inline andino_gazebo/urdf/andino.gazebo.xacro:57]: I see that you are not using the continuouts joint.
Does it make sense to have an argument in the launch file that you can pass down to the xacro for using or not using a fixed/continuous caster joint?

File: andino_gazebo/README.MD [hunk 5]
```diff
@@ -50,6 +50,13 @@ Finally, all the parameters of spawn an andino robot to put the robot in specifi
 ```
 ros2 launch andino_gazebo andino_one_robot.launch.py -s
 ```
+## Andino simulation with Gazebo ros2 control plugin
+
+```
+ros2 launch andino_gazebo andino_one_robot.launch.py use_gazebo_ros_control:=true
+```
+
+Some physics issue made that the robot do not move in straight-line with ros_control activated. Another limitation to take account is that the twist topic should be sent at a frequency of at least $10hz$ because it the action has a validity timeout. Finally, note that the twist topic is `diff_controller/cmd_vel_unstamped`.
```
Comment (francocipollone) [inline andino_gazebo/README.MD:59]: It looks good to me to provide this kind of information. Can you add this as bullets under a section `notes`?
e.g:
```
### Considerations
 - note that the twist topic is `diff_controller/cmd_vel_unstamped`.
 - phyisics issue made that ....
 - cmd topic should be sent at ...
 ```

**Review by francocipollone — CHANGES_REQUESTED** [review]
I will try this locally when I have time. I left some comments

File: andino_gazebo/README.MD (hunk 5 above)
Comment (agalbachicar) [reply andino_gazebo/README.MD:59]: For the record, I did not notice the effect you mentioned when sending different cmd vels. It is behaving just fine for me.

**Review by agalbachicar — CHANGES_REQUESTED** [review]
It looks good to me, just leaving a few comments for completion.

File: andino_description/urdf/include/andino_caster_macro.urdf.xacro (hunk 2 above)
Comment (olmerg) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: The problem is that the only solution that I found for ros control works  is with the caster_rotation_joint fixed and the caster_rotation_wheel continuous. I can isolate the error in one file is  the urdf is modified inside the spawn_launch, What do you think?

File: andino_gazebo/README.MD (hunk 5 above)
Comment (olmerg) [reply andino_gazebo/README.MD:59]: @agalbachicar I forget to delete this comment in the last commit

File: andino_description/urdf/include/andino_caster_macro.urdf.xacro (hunk 2 above)
Comment (gonzodepedro) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: Have you tried modelling the caster wheel collision geometry as a sphere instead of a cillinder? That would mean it will have a single point of contact with the floor, instead of a line of points.

Comment (BarceloChristian) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: Chiming in since I added the option to select the joint's type.
I understand different physics engines may behave differently and might require little tweaks as modeling the caster as a non-friction ball or a real-like implementation with joints.
What I don't understand is why that is a limitation for `ros_control`. You would never want the caster wheel to be controlled/sensed through `ros_control`, right? So why this is a problem?
I feel we're all asking and answering about different things, so knowing what the real problem is might be the best approach to discuss upon.

Comment (francocipollone) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: > The problem is that the only solution that I found for ros control works is with the caster_rotation_joint fixed and the caster_rotation_wheel continuous. I can isolate the error in one file is the urdf is modified inside the spawn_launch, What do you think?

My comment is focused on trying to not mess up the logic in the urdf description. With that change, the `use_fixed` arg lacks of consistency. The urdf description can be as flexible as is required to deal with simulator's issues while keeping the integrity of the description.

Comment (olmerg) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: I understand your approach , but the problem that I want to solve is that the robot do not move in straight-line with ros_control activated. Th only way that I have found is to emulate a sphere wheel with wheel friction zero ( caster_rotation_joint fixed and the caster_rotation_wheel continuous).

Comment (BarceloChristian) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: You don't need the caster to be attached to a continuous joint to define the wheel friction to zero. Keep it fixed (as it was before, both joints fixed) and set something along the lines of (not tested and written from the top of my head):
```
<gazebo reference="caster_wheel">
  <mu1>0.0</mu1>
  <mu2>0.0</mu2>
</gazebo>
```

Isn't that enough?

Comment (olmerg) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: this solution did not works I do not understand why, I have tested different combinations in the $\mu$ of the caster wheel and the motor wheels.

Comment (BarceloChristian) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: If you remove the friction from all the wheels, the robot wouldn't move at all.
Let me check on my own if I can make it work. Any suggestion on how to quickly bring it up and test it?

**Comment by francocipollone** [comment]
Is it working as expected? @olmerg . When using `use_gazebo_ros_control:=true` it fails to correctly open the simulation in my container. It works well in `false`.

File: andino_description/urdf/include/andino_caster_macro.urdf.xacro (hunk 2 above)
Comment (olmerg) [reply andino_description/urdf/include/andino_caster_macro.urdf.xacro:119]: following the readme should works

**Comment by olmerg** [comment]
Hi @francocipollone  I isolate the hack of the urdf in the spawn_lunch, what do you think?

**Comment by agalbachicar** [comment]
Fro the record, I tested it in 0ede169 and this is the result.

[Uploading andino_caster_test_0ede169.zip…]()

Don't worry about the right side. You can see the drift also Twist command is zero. This is certainly an issue with the modeling.

File: andino_gazebo/launch/spawn_robot.launch.py [hunk 6]
```diff
@@ -59,13 +59,22 @@ def get_robot_description(use_ros_control: str) -> str:
...
+            'use_real_ros_control': 'false',
+            'use_fixed_caster': 'false',
+        },
     )
     robot_desc = doc.toprettyxml(indent='  ')
     folder = get_package_share_directory('andino_description')
     robot_desc = robot_desc.replace(
         'package://andino_description/', f'file://{folder}/'
     )
+    # Solve the problem of the caster
```
Comment (francocipollone) [inline andino_gazebo/launch/spawn_robot.launch.py:73]: Can you explain the issue here? This replace is a workaround and as such it would be great if we can explain here why we are doing what we are doing. The alternative would be to create an issue in the repo and add the link here.

**Review by francocipollone — CHANGES_REQUESTED** [review]
LGTM. I left only one nit comment.
Holding approval because I can't use the ros control version locally yet.

**Comment by francocipollone** [comment]
> Fro the record, I tested it in [0ede169](https://github.com/Ekumen-OS/andino/commit/0ede169f1f62de4d8ffb3ae7ec21078679117b10) and this is the result.
> 
> Uploading andino_caster_test_0ede169.zip…
> 
> Don't worry about the right side. You can see the drift also Twist command is zero. This is certainly an issue with the modeling.

@agalbachicar 
Note that what you were uploading did not uploaded correclty (you probably hit comment before it finished uploading :nerd_face: )

**Review by francocipollone — APPROVED** [review]
LGTM! :rocket:

**Review by francocipollone — APPROVED** [review]

**Review by francocipollone — CHANGES_REQUESTED** [review]
There are changes that aren't correct. Please update your humble branch in your own fork

File: .gitignore [hunk 7]
```diff
@@ -1,2 +1,8 @@
 *.build
 *.install
+
+# PlatformIO files
+.pio
```
[the remark this answers is not in the document]
Comment (francocipollone) [reply .gitignore:5]: The merge is wrongly done. See the commit message.
